src/cache_entry.cc

```diff
@@ -0,0 +1,354 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "cache_entry.h"
+#include <iostream>
+
+namespace triton { namespace core {
+
+/* CacheEntry */
+
+size_t
+CacheEntry::ItemCount()
+{
+  std::unique_lock lk(item_mu_);
+  return items_.size();
+}
+
+std::vector<std::shared_ptr<CacheEntryItem>>
+CacheEntry::Items()
+{
+  std::unique_lock lk(item_mu_);
+  return items_;
+}
+
+void
+CacheEntry::AddItem(std::shared_ptr<CacheEntryItem> item)
+{
+  std::unique_lock lk(item_mu_);
+  // TODO: Look at this flow, ownership, etc. - Currently, cache needs to not
+  // delete/invalidate the item.
+  items_.push_back(item);
+}
+
+/* CacheEntryItem */
+
+size_t
+CacheEntryItem::BufferCount()
+{
+  std::unique_lock lk(buffer_mu_);
+  return buffers_.size();
+}
+
+std::vector<Buffer>
+CacheEntryItem::Buffers()
+{
+  std::unique_lock lk(buffer_mu_);
+  return buffers_;
+}
+
+void
+CacheEntryItem::AddBuffer(const void* base, size_t byte_size)
+{
+  std::unique_lock lk(buffer_mu_);
+  // COPY: Make a copy of buffer for Triton to own
+  void* new_base = malloc(byte_size);
```

Comment:
Do we need to user Triton's internal memory pool for this? CC @GuanLuo

Reply:
We can but I think the allocation is temporary as it looks like there is plan to avoid the copy
```
/// NOTE: (DLIS-4471) Currently this will make a copy of the buffer at 'base'
/// to avoid lifetime issues.
```